networkx/algorithms/centrality/domirank.py

```diff
@@ -0,0 +1,342 @@
+            )
+    psi = dict(zip(G, (psi).tolist()))
+    return psi, sigma, converged
+
+
+##### THE FUNCTIONS HEREUNDER SHOULD BE HIDDEN FUNCTIONS #####
+def _find_smallest_eigenvalue(
+    G,
+    min_val=0,
+    max_val=1,
+    max_depth=100,
+    dt=0.1,
+    epsilon=1e-5,
+    max_iter=100,
+    patience=10,
+):
+    """
+    This function is simply used to find the smallest eigenvalue, by seeing when the DomiRank algorithm diverges. It uses
+    a kind of binary search algorithm, however, with a bias to larger eigenvalues, as finding divergence is faster than
+    verifying convergence.
+    This function outputs the smallest eigenvalue - i.e. most negative eigenvalue.
+    """
+    import numpy
+    import scipy
+
+    x = (min_val + max_val) / G.sum(axis=-1).max()
+    for _ in range(max_depth):
+        if max_val - min_val < epsilon:
+            break
+        _, converged = _domirank_iterative(
+            G, sigma=x, dt=dt, epsilon=epsilon, max_iter=max_iter, patience=patience
+        )
+        if converged:
+            min_val = x
+        else:
+            max_val = (x + max_val) / 2
+        x = (min_val + max_val) / 2
+    final_val = (max_val + min_val) / 2
+    return -1 / final_val
+
+
+def _domirank_iterative(GAdj, sigma=0, dt=0.1, epsilon=1e-5, max_iter=100, patience=10):
+    """
+    This function is used for the iterative computation of domirank when the comp_method = "iterative".
+    It is also used to find the smallest eigenvalue - i.e. called in the _find_smallest_eigenvalue() function.
+
+    Parameters
+    ----------
+    G : graph
+        A NetworkX graph.
+
+    sigma: float, optional (default=0.95)
+        The normalized level of competition for DomiRank.
+
+    dt: float, optional (default=0.1)
+        The step size for the Newton iteration.
+
+    epsilon: float, optional (default=1e-5)
+        The relative stopping criterion for convergence.
+
+    max_iter: integer, optional (default=100)
+        Maximum number of Newton iterations allowed.
+        It is recommended that ''max_iter >= 50''.
+
+    patience: integer, optional (default=10)
+        The number of steps between convergence checks.
+        It is recommended that ''patience >= 10''.
+
+
+    Returns
+    -------
+    psi : np.array
+        An array of the DomiRank values ordered according to the input Adjacency matrix.
+
+    converged : boolean
+        Whether the centrality computation converged.
+    """
+    import numpy as np
+    import scipy as sp
+
+    # store this to prevent more redundant calculations in the future
+    pGAdj = sigma * GAdj.astype(np.float32)
+    # initalize a proportionally (to system size) small non-zero uniform vector
+    psi = np.ones(pGAdj.shape[0], dtype=np.float32) / pGAdj.shape[0]
+    # initialize a zero array to store values (yes, this could be done with a smaller array with some smart indexing, but this is not computationally or memory heavy)
```

Comment:
```suggestion
    # initialize a zero array to store values (this could be done with a smaller array with some smart indexing, but isn't computationally or memory heavy)
```